compiler/rustc_serialize/src/opaque.rs

```diff
@@ -38,210 +35,97 @@ pub struct FileEncoder {
 
 impl FileEncoder {
     pub fn new<P: AsRef<Path>>(path: P) -> io::Result<Self> {
-        // Create the file for reading and writing, because some encoders do both
-        // (e.g. the metadata encoder when -Zmeta-stats is enabled)
-        let file = File::options().read(true).write(true).create(true).truncate(true).open(path)?;
-
         Ok(FileEncoder {
-            buf: Box::new_uninit_slice(BUF_SIZE),
+            buf: vec![0u8; BUF_SIZE].into_boxed_slice().try_into().unwrap(),
             buffered: 0,
             flushed: 0,
-            file,
+            file: File::create(path)?,
             res: Ok(()),
         })
     }
 
     #[inline]
     pub fn position(&self) -> usize {
         // Tracking position this way instead of having a `self.position` field
-        // means that we don't have to update the position on every write call.
+        // means that we only need to update `self.buffered` on a write call,
+        // as opposed to updating `self.position` and `self.buffered`.
         self.flushed + self.buffered
     }
 
-    pub fn flush(&mut self) {
-        // This is basically a copy of `BufWriter::flush`. If `BufWriter` ever
-        // offers a raw buffer access API, we can use it, and remove this.
-
-        /// Helper struct to ensure the buffer is updated after all the writes
-        /// are complete. It tracks the number of written bytes and drains them
-        /// all from the front of the buffer when dropped.
-        struct BufGuard<'a> {
```

Comment:
Why is the old version of `flush` so complicated?...

Reply:
I think approximately all of the complexity of the previous implementation was due to pasting in `BufWriter` then gradually hacking it up to support direct buffer access. `BufWriter` cares about being able to back out of a failed write so that the user could attempt to flush it again or something like that. But `FileEncoder` doesn't.